### PR TITLE
add channels last support for PixelShuffle and PixelUnshuffle

### PR DESCRIPTION
**Pull request by mingfeima**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #50573

Differential Revision: [D26007228](https://our.internmc.facebook.com/intern/diff/D26007228)

### CODE REVIEW

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -2805,8 +2805,24 @@
 
 - func: pixel_shuffle(Tensor self, int upscale_factor) -> Tensor
 
+- func: _pixel_shuffle_cpu(Tensor self, int[] output_sizes, int upscale_factor) -> Tensor
+  dispatch:
+    CPU: _pixel_shuffle_cpu
```
Comment (ezyang) [inline aten/src/ATen/native/native_functions.yaml:2810]: This looks weird. Why can't we just modify the existing pixel shuffle to look like this:

```
- func: pixel_shuffle(Tensor self, int upscale_factor) -> Tensor
  dispatch:
   CompositeImplicitAutograd: pixel_shuffle
   CPU: pixel_shuffle_cpu 
```

If you are concerned about having to recompute `output_sizes` in both implementations, consider making this a structured kernel: https://github.com/pytorch/rfcs/blob/rfc-0005/RFC-0005-structured-kernel-definitions.md

Comment (mingfeima) [reply aten/src/ATen/native/native_functions.yaml:2810]: COOL! Thanks!
I wasn't aware that pytorch has this `CompositeImplicitAutograd` dispatch key......

**Comment by ezyang** [comment]
I don't think I'm the primary reviewer on this, but it doesn't seem like there has been any reviewer action. @mingfeima do you know who should be reviewing this?

File: aten/src/ATen/native/native_functions.yaml (hunk 1 above)
Comment (mingfeima) [reply aten/src/ATen/native/native_functions.yaml]: updated with `CompositeImplicitAutograd` dispatch key!

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D26007228).

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D26007228).

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D26007228).

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D26007228).

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D26007228).

**Comment by mingfeima** [comment]
TODO：
- [ ] add benchmark
- [ ] add dtype Byte

**Comment by mingfeima** [comment]
i run into the following error in CI, could you please give some suggestions @ezyang , thanks!

```bash
The PR is introducing backward incompatible changes to the operator library. Please contact PyTorch team to confirm whether this change is wanted or not. 

Broken ops: [
	aten::_test_warn_in_autograd(Tensor self) -> (Tensor)
	aten::matrix_H(Tensor(a) self) -> (Tensor(a))
	aten::matrix_H.a(Tensor(a) self) -> (Tensor(a))
	aten::mT(Tensor(a) self) -> (Tensor(a))
	aten::mT.a(Tensor(a) self) -> (Tensor(a))
	aten::numpy_T.a(Tensor(a) self) -> (Tensor(a))
	aten::adjoint(Tensor(a) self) -> (Tensor(a))
	aten::grid_sampler_2d_backward(Tensor grad_output, Tensor input, Tensor grid, int interpolation_mode, int padding_mode, bool align_corners, bool[2] output_mask) -> (Tensor, Tensor)
	aten::mH(Tensor(a) self) -> (Tensor(a))
	aten::mH.a(Tensor(a) self) -> (Tensor(a))
]
```

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D26007228).

**Review by VitalyFedyunin — APPROVED** [review]

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D26007228).

**Comment by mingfeima** [comment]
@VitalyFedyunin  the unused variables from  `cpu/PixelShuffleKernel.cpp` has been removed, please review!

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D26007228).

**Comment by VitalyFedyunin** [comment]
Merge conflicts with master, can you please rebase :(

**Comment by mingfeima** [comment]
This stack has recently been rebased!

**Comment by mingfeima** [comment]
This stack has been rebased recently!

**Comment by mingfeima** [comment]
Rebased!

**Comment by mingfeima** [comment]
@VitalyFedyunin The stack is newly rebased, please review~

**Comment by mingfeima** [comment]
fix failures from `test_public_bindings.py`

File: test/allowlist_for_publicAPI.json [hunk 2]
```diff
@@ -2404,6 +2404,8 @@
     "linear",
     "logsigmoid",
     "native_channel_shuffle",
+    "native_pixel_shuffle",
```
Comment (anjali411) [inline test/allowlist_for_publicAPI.json:2407]: please undo the changes in this file. This should be fixed by https://github.com/pytorch/pytorch/pull/75801

**Comment by anjali411** [comment]
@mingfeima I can merge this PR after you rebase and revert the changes in `test/allowlist_for_publicAPI.json` :)

File: test/allowlist_for_publicAPI.json (hunk 2 above)
Comment (mingfeima) [reply test/allowlist_for_publicAPI.json]: @anjali411 thanks for reminding! This change is removed!

**Comment by mingfeima** [comment]
@anjali411 still met the following failures, any suggestions? thanks
```
FAIL [1.177s]: test_correct_module_names (__main__.TestPublicBindings)
An API is considered public, if  its  `__module__` starts with `torch.`
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_public_bindings.py", line 350, in test_correct_module_names
    self.assertTrue(not failure_list, msg)
AssertionError: False is not true : Following new APIs ( displayed in the form (module, element, element module) ) were added that do not meet our guidelines for public API Please review  https://docs.google.com/document/d/10yx2-4gs0gTMOimVS403MnoAWkqitS8TUHX73PN8EjE/edit?pli=1# for more information:
('torch.nn.functional', 'native_pixel_shuffle', 'torch')
('torch.nn.functional', 'native_pixel_unshuffle', 'torch')
```

**Comment by mingfeima** [comment]
abstract this one from the main channels last PR stack.

**Comment by mingfeima** [comment]
the CI failures are cleared, please review!

File: torch/nn/functional.py [hunk 3]
```diff
@@ -3512,6 +3512,60 @@ def multi_margin_loss(
 """,
 )
 
+_native_pixel_shuffle = _add_docstr(
+    torch._native_pixel_shuffle,
+    r"""
+_native_pixel_shuffle(input, upscale_factor) -> Tensor
+
+Native kernel level implementation of the `pixel_shuffle`.
+This function might become private in future releases, use with caution.
```
Comment (jbschlosser) [inline torch/nn/functional.py:3521]: doc nit: the function is private now since it begins with an underscore

File: torch/nn/functional.py [hunk 4]
```diff
@@ -3512,6 +3512,60 @@ def multi_margin_loss(
...
+""",
+)
+
+_native_pixel_unshuffle = _add_docstr(
+    torch._native_pixel_unshuffle,
+    r"""
+_native_pixel_unshuffle(input, downscale_factor) -> Tensor
+
+Native kernel level implementation of the `pixel_unshuffle`.
+This function might become private in future releases, use with caution.
```
Comment (jbschlosser) [inline torch/nn/functional.py:3547]: doc nit: the function is private now since it begins with an underscore

File: aten/src/ATen/native/native_functions.yaml [hunk 5]
```diff
@@ -3445,8 +3445,18 @@
...
+    CPU: pixel_shuffle_cpu
+    CompositeImplicitAutograd: math_pixel_shuffle
+
 - func: pixel_unshuffle(Tensor self, int downscale_factor) -> Tensor
 
+- func: _native_pixel_unshuffle(Tensor self, int downscale_factor) -> Tensor
+  dispatch:
+    CPU: pixel_unshuffle_cpu
+    CompositeImplicitAutograd: math_pixel_unshuffle
+
```
Comment (jbschlosser) [inline aten/src/ATen/native/native_functions.yaml:3459]: Why introduce `_native_pixel_shuffle` / `_native_pixel_unshuffle` rather than just change the dispatch key entries for `pixel_shuffle` / `pixel_unshuffle` directly?

**Review by jbschlosser — COMMENTED** [review]
Made a quick pass with some comments :)

File: torch/nn/functional.py (hunk 3 above)
Comment (mingfeima) [reply torch/nn/functional.py]: This issue has been fixed, remove `_native_pixel_shuffle`.

File: aten/src/ATen/native/native_functions.yaml (hunk 5 above)
Comment (mingfeima) [reply aten/src/ATen/native/native_functions.yaml]: remove `_native_pixel_shuffle` and `_native_pixel_unshuffle`

File: torch/nn/functional.py (hunk 4 above)
Comment (mingfeima) [reply torch/nn/functional.py]: fixed

**Comment by mingfeima** [comment]
Test failures have been removed, could you please help review! @VitalyFedyunin @anjali411 @jbschlosser

**Comment by mingfeima** [comment]
rebased to trigger CI.

**Comment by VitalyFedyunin** [comment]
Please rebase again, having troubles importing it.

**Comment by mingfeima** [comment]
Rebased to latest! Please review!

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D26007228).
